### PR TITLE
RATIS-1943. Pending read requests instead of failing when leader not ready

### PR DESCRIPTION
**Pull request by SzyWilliam**
See https://issues.apache.org/jira/browse/RATIS-1943.

When leader is elected but not ready, we should pending the incoming read-index requests and reply later, instead of failing requests immediately.

### CODE REVIEW

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 1]
```diff
@@ -356,6 +357,7 @@ boolean isApplied(LogEntryProto logEntry) {
   private final PendingStepDown pendingStepDown;
 
   private final ReadIndexHeartbeats readIndexHeartbeats;
+  private final ConcurrentLinkedQueue<CompletableFuture<Long>> pendingReadIndexRequests;
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:360]: We may reuse `startupLogEntry`.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 2]
```diff
@@ -424,6 +427,7 @@ void checkReady(LogEntryProto entry) {
     Preconditions.assertTrue(startupLogEntry.isInitialized());
     if (entry.getTerm() == getCurrentTerm() && startupLogEntry.get().isApplied(entry)) {
       server.getStateMachine().leaderEvent().notifyLeaderReady();
+      releasePendingRequests(entry.getIndex(), null);
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:430]: `startupLogEntry.get().isApplied(..)` above will take care this.

**Review by szetszwo — COMMENTED** [review]
@SzyWilliam , thanks for working on this!  The change is surprisingly simple if we reuse `startupLogEntry`; see below
```java
diff --git a/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java b/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java
index e1b12d48..2465fdc2 100644
--- a/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java
+++ b/ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java
@@ -306,6 +305,10 @@ class LeaderStateImpl implements LeaderState {
     /** This future will be completed after the log entry is applied. */
     private final CompletableFuture<Long> appliedIndexFuture = new CompletableFuture<>();
 
+    CompletableFuture<Long> getAppliedIndexFuture() {
+      return appliedIndexFuture;
+    }
+
     boolean isApplied(LogEntryProto logEntry) {
       if (appliedIndexFuture.isDone()) {
         return true;
@@ -444,9 +447,9 @@ class LeaderStateImpl implements LeaderState {
       LOG.warn("{}: Caught exception in sendNotLeaderResponses", this, e);
     }
     messageStreamRequests.clear();
-    // TODO client should retry on NotLeaderException
     readIndexHeartbeats.failListeners(nle);
     lease.getAndSetEnabled(false);
+    startupLogEntry.get().getAppliedIndexFuture().completeExceptionally(nle);
     server.getServerRpc().notifyNotLeader(server.getMemberId().getGroupId());
     logAppenderMetrics.unregister();
     raftServerMetrics.unregister();
@@ -1117,12 +1120,9 @@ class LeaderStateImpl implements LeaderState {
       return CompletableFuture.completedFuture(readIndex);
     }
 
-    // leader has not committed any entries in this term, reject
-    // TODO: wait for leader to become ready instead of failing the request.
+    // leader has not committed any entries in this term
     if (!isReady()) {
-      return JavaUtils.completeExceptionally(new ReadIndexException(
-          "Failed to getReadIndex " + readIndex + " since the term is not yet committed.",
-          new LeaderNotReadyException(server.getMemberId())));
+      return startupLogEntry.get().getAppliedIndexFuture();
     }
 
     // if lease is enabled, check lease first
```

**Comment by SzyWilliam** [comment]
@szetszwo You are right! The change looks simple and elegant! Thanks very much for the suggestions! There're always surprises when reading the Ratis source code ;)

BTW, I think we need to throw a `ReadIndexException` instead of `NotLeaderException` to be compatible with original client side. Let me just wrap the `nle` with a `ReadIndexException`

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**